### ISSUE TITLE
Question: How to use a phantom type to disambiguate which interpreter to use?

### ISSUE THREAD

**Issue by csasarak** [isovector/polysemy#45]
I am trying to create a custom effect which handles going to a database. Here is a minimal example:
```haskell
-- select returning an 'a'
data Select a = Select a

data DBAction whichDb m a where
  DoSelect :: Select a -> DBAction whichDb m (Maybe a)

makeSem ''DBAction
```
The idea is that one effect interpreter will maintain a pool of connections and these actions will be reduced to actual IO actions against them. However, in a program that works with multiple databases there needs to be some way to disambiguate which pool of connections to use for each query. I added the `whichDb` type variable to describe this and get this error:
```haskell
    • Ambiguous use of effect 'DBAction'
      Possible fix:
        add (Member (DBAction whichDb0) r) to the context of 
          the type signature
      If you already have the constraint you want, instead
        add a type application to specify
          'whichDb0' directly
    • In the ambiguity check for ‘doSelect’
      To defer the ambiguity check to use sites, enable AllowAmbiguousTypes
      In the type signature:
        doSelect :: forall a_Xfbf (whichDb_afbf :: k_afpj).
                    forall r_afXj.
                    Member (DBAction whichDb_afbf) r_afXj =>
                    Select a_Xfbf -> Sem r_afXj (Maybe a_Xfbf)
   |
84 | makeSem ''DBAction
   | ^^^^^^^^^^^^^^^^^^
```

Turning on polysemy-plugin makes that go away, but then a similar message pops up at the use sites, even with a TypeApplication. I can make it go away as well by putting `whichDb` as an argument to the data constructor, which is probably OK because it'll mean no more work for users (I say users, but it's just me) of `doSelect` than adding a type application.

I'm wondering if I've organized the effect wrong (so this is just PEBKAC) or if there's something to be done with polysemy. I am willing to help, would probably just need a little hand-holding - I am only just learning the type-level stuff.

Somewhat related, maybe it would be useful to add an explanation in the haddocks for exactly what code `makeSem` generates and why. Again, I am willing to help as best I can - thank you for making a very usable library!

**Comment by isovector** [comment]
Yeah, that's a known bug in the TH-generation --- phantom parameters aren't properly assigned in the smart constructors.

The [existing documentation](http://hackage.haskell.org/package/polysemy-0.1.2.0/docs/Polysemy.html#g:5) is hopefully enough to show you what's happening behind the TH, and in the future, you can always turn on `-ddump-splices` to see what code GHC is generating behind the scenes.

What do your call sites look like? It's weird to me that neither the plugin nor a type-application would fix the issue.

**Comment by csasarak** [comment]
So, I checked again and was mistaken, the minimal example doesn't show everything off. I'm still not sure why a manual type application didn't work. 

Here is a basic interpreter:

```haskell
runDBAction :: Sem (DBAction whichDb ': r) a -> Sem r a
runDBAction = interpret $ \(DoSelect (Select b)) -> return $ Just b
```
and use:
```haskell
testEffect :: (Member (DBAction Warehouse) r) => Sem r Integer
testEffect = do res <- doSelect @Warehouse (Select 10 :: Select Integer)
                return $ maybe 0 fromIntegral res
```
and I get this message:

```haskell
/Users/chrissasarak/Work/rhino-watcher/src/Watcher/Server.hs:62:24: error:
    • Ambiguous use of effect 'DBAction'
      Possible fix:
        add (Member (DBAction whichDb0) r) to the context of 
          the type signature
      If you already have the constraint you want, instead
        add a type application to specify
          'whichDb0' directly
    • In a stmt of a 'do' block:
        res <- doSelect @Warehouse (Select 10 :: Select Integer)
      In the expression:
        do res <- doSelect @Warehouse (Select 10 :: Select Integer)
           return $ maybe 0 fromIntegral res
      In an equation for ‘testEffect’:
          testEffect
            = do res <- doSelect @Warehouse (Select 10 :: Select Integer)
                 return $ maybe 0 fromIntegral res
   |
62 | testEffect = do res <- doSelect @Warehouse (Select 10 :: Select Integer)
```
It's worth noting though, and I just noticed this, that polysemy-plugin is turned on in the file where makeSem is but not enabled for `testEffect`, which is in a separate module.

However, when I do turn on `polysemy-plugin` I get a separate error, which I'm guessing is because of some type I wrote but I don't quite grok it yet.

```haskell
/Users/chrissasarak/Work/rhino-watcher/src/Watcher/Server.hs:1:1: error:
    solveSimpleWanteds: too many iterations (limit = 4)
      Set limit with -fconstraint-solver-iterations=n; n=0 for no limit
      Simples = {[WD] $dMonad_alMu {0}:: Monad m0 (CNonCanonical),
                 [WD] $d(%,,%)_alMB {0}:: Member
                                            (DBAction whichDb0) r0 (CNonCanonical),
                 [WD] $dNum_alMK {0}:: Num a0 (CNonCanonical),
                 [WD] $dMonad_alMO {0}:: Monad m0 (CNonCanonical),
                 [WD] $dNum_alMV {0}:: Num a0 (CNonCanonical),
                 [WD] $dIntegral_alNc {0}:: Integral a0 (CNonCanonical),
                 [WD] $dNum_alNd {0}:: Num b0 (CNonCanonical)}
      WC = WC {wc_simple =
                 [WD] hole{co_amwG} {1}:: DBAction whichDb0
                                          ~ DBAction Warehouse (CNonCanonical)
                 [WD] hole{co_amwu} {1}:: whichDb0 ~~ Warehouse (CIrredCan(sol))
                 [WD] hole{co_amwt} {1}:: Warehouse ~ * (CIrredCan(insol))
                 [WD] hole{co_amwf} {1}:: Warehouse ~ * (CIrredCan(insol))
                 [WD] hole{co_amw1} {1}:: Warehouse ~ * (CIrredCan(insol))
                 [W] irred_amwE {1}:: Polysemy.Internal.CustomErrors.Break
                                        (TypeError ...)
                                        (Polysemy.Internal.Union.IndexOf
                                           r
                                           (Polysemy.Internal.Union.Found
                                              r (DBAction whichDb0))) (CIrredCan(sol))
                 [WD] $dFind_amvT {1}:: Polysemy.Internal.Union.Find
                                          r (DBAction whichDb0) (CDictCan(psc))
                 [WD] hole{co_amwz} {7}:: Polysemy.Internal.Union.IndexOf
                                            r (Polysemy.Internal.Union.Found r (DBAction whichDb0))
                                          ~ DBAction whichDb0 (CNonCanonical)
                 [D] _ {1}:: Warehouse ~ * (CIrredCan(insol))
                 [D] _ {1}:: Warehouse ~ * (CIrredCan(insol))
                 [D] _ {1}:: Warehouse ~ * (CIrredCan(insol))
                 [D] _ {1}:: Warehouse ~ * (CIrredCan(insol))
                 [D] _ {1}:: Warehouse ~ * (CIrredCan(insol))
                 [D] _ {1}:: Polysemy.Internal.CustomErrors.Break
                               (TypeError ...)
                               (Polysemy.Internal.Union.IndexOf
                                  r (Polysemy.Internal.Union.Found r (DBAction whichDb0)))
                             ~ () :: Constraint (CNonCanonical)}
  |
1 | {-# OPTIONS_GHC -fplugin=Polysemy.Plugin #-}
  | ^
```

In any case, I'm guessing that when/if the TH bug is fixed I would be able to write this code without polysemy-plugin if I wanted? Or is it a hard requirement of the library? (besides the TH thing)

**Comment by isovector** [comment]
I think what's happening:

* Because `whichDb` is phantom, it receives a polykind `k`. 
* The kind of `DBAction` is thus `DBAction :: forall k -> k -> forall m -> m -> * -> *`, where the `forall k -> k`  bit comes from `whichDb` and the `forall m -> m` stuff is from `m` (which also has a polykind, but is expected by polysemy). You can use `-fprint-explicit-kinds` and `-fprint-explicit-foralls` to verify this is the kind `DBAction` gets.
* This kind is invisible and usually passed around by GHC without any problems.
* However, some interaction with the plugin attempts to fill in the *invisible* `forall k` parameter instead of the desired `k` parameter. 
* This leads to the `too many iterations` problem, since it keeps trying to generate a shitty wanted constraint.
* In fact, the plugin should fill in both the `forall k` and the `k` parameters.

In short, I suspect your phantom `whichDb` parameter is causing two separate pieces of grief! If all of this conjecture is correct, you should be able to fix it by putting an explicit kind on `whichDb`:

```haskell
data DBAction (whichDb :: *) m a where
```

(or `Symbol` or what have you)

**Comment by isovector** [comment]
To be clear, I think there are two bugs here, neither of which is your fault:

* `polysemy`'s TH generates ambiguous smart constructors for phantom parameters
* `polysemy-plugin` shits the bed when dealing with polykinds

**Comment by csasarak** [comment]
Ok so those don't work exactly, but I was able to try these scenarios:

- Turning on `polysemy-plugin` on where the effect is defined, and also specifying a kind in the GADT  results in the same error about ambiguity
- Keeping `polysemy-plugin` on while also defining the smart constructor manually works, and the whole program compiles.
- Defining the smart constructor manually with `polysemy-plugin` off results in the ambiguity error, even with a specific kind.

For reference, this is my manual version:

```haskell
doSelect :: (Member (DBAction whichDb) r) => Select a -> Sem r (Maybe a)
doSelect s = Polysemy.Internal.send (DoSelect s)
```
and this is the `makeSem` version:
```haskell
    type instance Polysemy.Internal.CustomErrors.DefiningModule DBAction = "Watcher.DBSupport"
    doSelect ::
      forall a_XffT (whichDb_affT :: k_aftV).
      forall r_ag1V.
      Member (DBAction whichDb_affT) r_ag1V =>
      Select a_XffT -> Sem r_ag1V (Maybe a_XffT)
    {-# INLINABLE CONLIKE doSelect #-}
    doSelect a_ag1W = Polysemy.Internal.send (DoSelect a_ag1W)
```

I think I should be able to proceed, but please let me know if I can provide more information.

**Comment by isovector** [comment]
Try this:

```haskell
doSelect :: forall whichDb a r. (Member (DBAction whichDb) r) => Select a -> Sem r (Maybe a)
doSelect s = Polysemy.Internal.send (DoSelect @whichDb s)
```

which should fix the ambiguity errors at the definition site. 

Do you still get the gnarly `too many iterations` error with the manual kind signature on the defition of `DBAction`?

**Comment by csasarak** [comment]
Just because it's spread over many comments, I'm working with these definitions now, the only change I made to your `doSelect` is using `@_` to shift the `whichDb` application over, otherwise it looked like it was trying to prove `a ~ whichDb`:
```haskell
data Select a = Select a

data DBAction (whichDb :: *) m a where
  DoSelect :: Select a -> DBAction whichDb m (Maybe a)

doSelect :: forall whichDb a r. (Member (DBAction whichDb) r) => Select a -> Sem r (Maybe a)
doSelect s = Polysemy.Internal.send (DoSelect @_ @whichDb s)

runDBAction :: Sem (DBAction whichDb ': r) a -> Sem r a
runDBAction = interpret $ \(DoSelect (Select b)) -> return $ Just b

```
With `polysemy-plugin` it works fine, both the definition and the use-site compile.

Using that manual definition without `polysemy-plugin` I get the ambiguity error still. The compiler points to the type definition itself:

```haskell
    • Ambiguous use of effect 'DBAction'
      Possible fix:
        add (Member (DBAction whichDb0) r) to the context of 
          the type signature
      If you already have the constraint you want, instead
        add a type application to specify
          'whichDb0' directly
    • In the ambiguity check for ‘doSelect’
      To defer the ambiguity check to use sites, enable AllowAmbiguousTypes
      In the type signature:
        doSelect :: forall whichDb a r.
                    (Member (DBAction whichDb) r) => Select a -> Sem r (Maybe a)
   |
87 | doSelect :: forall whichDb a r. (Member (DBAction whichDb) r) => Select a -> Sem r (Maybe a)
   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

I don't get any errors about `too many iterations` though!

**Comment by isovector** [comment]
Thanks for the clarification! It looks like you're unstuck now after enabling the plugin everywhere?

**Comment by csasarak** [comment]
I am, I'd like to understand why type applications didn't fix the issue but I can definitely move forward. Thank you!

**Comment by isovector** [comment]
I'm not sure either, though I think my official recommendation is just "always use the plugin!" :)

I'm going to close this issue, but spin up two for the bugs we uncovered here.
